(1.20.1) Adds Capability for Mods to Register Custom Brewing Fuels

Thomas1034 opened the pull request:
It is currently impossible to make brewing stands accept alternate fuels in 1.20.1. This pull request fixes that problem. It adds a new class, `BrewingFuelRegistry`, which allows mods to associate an item with a brewing stand fuel value. Brewing fuels - like brewing recipes - can be added to the registry during the `FMLCommonSetupEvent` event.

I have also confirmed that this works with hoppers; they can correctly insert new fuels.

A test mod is included, verifying that both custom items and vanilla items can be added as fuels. 

Please let me know if there are any problems I should address with this pull request.

Thomas1034 committed:
Create HangingSignBlockEntity.java.patch

Added a constructor that takes in a block entity type. This allows mods to create their own hanging sign block entities without needing to implement their own rendering methods for the GUI, use mixins, or attempt to hardcode the getter methods. See net.minecraft.client.player$openTextEdit(net.minecraft.world.level.block.entity.SignBlockEntity, boolean) for the reason this would be useful - to render the GUI for a hanging sign, the block entity needs to extend HangingSignBlockEntity.

Thomas1034 committed:
Update HangingSignBlockEntity.java.patch

Thomas1034 committed:
Update HangingSignBlockEntity.java.patch

Thomas1034 committed:
Adding Alternate Brewing Fuels

Partially done.

Thomas1034 committed:
Relocated

Thomas1034 committed:
Merge branch '1.20.1-extensible-hanging-sign-be' of https://github.com/Thomas1034/MinecraftForge into 1.20.1-extensible-hanging-sign-be

Thomas1034 committed:
Merge pull request #1 from MinecraftForge/1.20.1

1.20.1 - Import Changes

Thomas1034 committed:
Merge branch '1.20.1' into 1.20.1-brewing-fuels

Thomas1034 committed:
Fixed rearrangements to ForgeHooks

Also removed extra line in the event class.

Thomas1034 committed:
Add import to ForgeHooks

Thomas1034 committed:
Generated patches + added jsons

Thomas1034 committed:
Update ConstructorThrowTest.java

Fixed accidental comment out

Thomas1034 committed:
Fixed bizarre glitch. I hope.

Why won't GitHub work right?

Thomas1034 committed:
Spacing

Thomas1034 committed:
Slight renaming, changed stack size of item

Thomas1034 committed:
Merge pull request #2 from Thomas1034/1.20.1-brewing-fuels

(1.20.1) Adds Capability for Mods to Register Custom Brewing

LexManos commented on src/main/java/net/minecraftforge/event/brewing/RegisterBrewingFuelEvent.java:
```diff
@@ -0,0 +1,30 @@
+
+public class RegisterBrewingFuelEvent extends net.minecraftforge.eventbus.api.Event 
+{
+    
```
Our new code style is to have braces on the same line, so please do that.

Thomas1034 replied:
Understood. I was wondering if the "how to contribute to forge" document I found might be a little out of date... guess so. It's [here](https://github.com/MinecraftForge/MinecraftForge/wiki/If-you-want-to-contribute-to-Forge) if you want to take a look.

LexManos commented on patches/minecraft/net/minecraft/world/inventory/BrewingStandMenu.java.patch:
```diff
@@ -1,5 +1,13 @@
+ import net.minecraft.world.item.alchemy.PotionUtils;
+ 
+ public class BrewingStandMenu extends AbstractContainerMenu {
++   private static final net.minecraftforge.common.BrewingFuelValues BREWING_FUELS = net.minecraftforge.common.ForgeHooks.onBrewingFuelRegisterEvent();
```
This fires the event during the static initializer of the brewing stand menu, this is not a good place for that. A better place would be to make it part of the registry events, Or perhaps instead of making it a custom registry implementation. It could be data driven and use the syncing mechanics of the custom data registries?

Thomas1034 replied:
I'll look into those and see if I can find a better way. One thing, though: you mentioned [on Discord](https://discord.com/channels/1129059589325852724/1129069883813007360/1293357425344122930) that using tags would not be the best idea - wouldn't making it data driven have the same problems with vanilla server/Forge client?

LexManos replied:
Quite possibly, it would need testing and a way to verify that connecting to vanilla servers behaves fine.

Thomas1034 replied:
I'm having trouble figuring out how to use the custom data registries. Are there any existing examples I could take a look at? 

Since I've been so far unable to figure that part out, I've been working on a different approach. I reworked the brewing fuel values class to pattern it off of the `BrewingRecipeRegistry` class; thus, instead of having its own event, mods can add values to a static map in the `FMLCommonSetupEvent`. This version does not require adding an event, requires only minimal changes to other Forge classes, and matches the existing way of adding behaviors to the brewing stand.

LexManos commented on patches/minecraft/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java.patch:
```diff
@@ -1,5 +1,24 @@
+ import net.minecraft.world.level.block.state.BlockState;
+ 
+ public class BrewingStandBlockEntity extends BaseContainerBlockEntity implements WorldlyContainer {
++   private static final net.minecraftforge.common.BrewingFuelValues BREWING_FUELS = net.minecraftforge.common.ForgeHooks.onBrewingFuelRegisterEvent();
```
This again has the static initializer issue. But also you're firing the event twice, once here and once in the menu. Better is to find a way to fire it once and use it in both places.

Thomas1034 replied:
I resolved this by making `BrewingFuelValues` work in a static way, like how `BrewingRecipeRegistry` does. This should make things simpler for both players and maintenance.

LexManos commented on src/main/java/net/minecraftforge/common/BrewingFuelValues.java:
```diff
@@ -0,0 +1,61 @@
+
+    public BrewingFuelValues() 
+    {
+        this.fuelTimes = new Object2IntOpenHashMap<>();
```
As this takes no values from the constructor it can be done in the field initializer.

LexManos commented on src/main/java/net/minecraftforge/common/BrewingFuelValues.java:
```diff
@@ -0,0 +1,61 @@
+     */
+    public int get(Item fuel) 
+    {
+        return this.fuelTimes.getOrDefault(fuel, VANILLA_FUEL_TIME);
```
This would make any item not registered return the default. I dont think this is good api...

RealMangorage replied:
The idea was that items tagged as brewing stand fuel would be whatever the default is, and can be overriden via this method.

LexManos replied:
Yes, but that is not what this method does, thats what the add(Item) overload does.
Change this to `getOrDefault(fule, 0)` and get rid of `hasFuel` in favor of `get(item) != 0`

Thomas1034 replied:
Understood, I'll fix that. Thank you for your input.

Thomas1034 committed:
Updated to address concerns.

Also drastically simplified the way it worked, and removed the (now unnecessary) event and hook.

Thomas1034 committed:
Merge pull request #3 from Thomas1034/1.20.1-brewing-fuels

Updated to address concerns.

Thomas1034 committed:
Update ConstructorThrowTest.java

Thomas1034 committed:
Merge pull request #4 from Thomas1034/1.20.1-brewing-fuels

Update ConstructorThrowTest.java (Oops)

Thomas1034 commented:
I've updated the pull request to address the concerns mentioned above.

Thomas1034 committed:
Fixed Patches

In the previous commit, patches were not generated correctly; apparently, it took in excess of five minutes for Git to realize the patch files had changed after the task had finished, so the previous commit did not include them.

Thomas1034 committed:
Merge pull request #5 from Thomas1034/1.20.1-brewing-fuels

Fixed Patches

Jonathing reviewed (changes requested):
Please update the PR description to account for the changes made in recent commits.

Thomas1034 commented:
I have updated the description as [Jonathing](https://github.com/Jonathing) requested.

Jonathing commented on patches/minecraft/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java.patch:
```diff
@@ -1,5 +1,16 @@
+ 
+    public static void m_155285_(Level p_155286_, BlockPos p_155287_, BlockState p_155288_, BrewingStandBlockEntity p_155289_) {
+       ItemStack itemstack = p_155289_.f_58975_.get(4);
+-      if (p_155289_.f_58979_ <= 0 && itemstack.m_150930_(Items.f_42593_)) {
+-         p_155289_.f_58979_ = 20;
++      if (p_155289_.f_58979_ <= 0 && (itemstack.m_150930_(Items.f_42593_) || net.minecraftforge.common.BrewingFuelRegistry.get(itemstack.m_41720_()) != 0)) {
++         p_155289_.f_58979_ = net.minecraftforge.common.BrewingFuelRegistry.get(itemstack.m_41720_());
```
You should move `BrewingFuelRegistry.get()` to its own variable. This will avoid calling it twice, along with the parameters which also include method calls.

Thomas1034 replied:
I've fixed that. I'm not sure why I didn't have it do that originally.

Jonathing commented on src/main/java/net/minecraftforge/common/BrewingFuelRegistry.java:
Please add a way to clarify in this file that you can only add to this once `FMLCommonSetup` has been fired.

Thomas1034 replied:
I've updated the comments to indicate that it should only be manipulated during the `FMLCommonSetupEvent`.

Jonathing reviewed (changes requested):
I love this PR. Incredibly clear, efficient, and does what it needs to do well. I have a few nitpicks, but after that I think it's good for merge. I will handle backports for relevant versions.

Thomas1034 committed:
Updated comments and moved get() into variable

Thomas1034 committed:
Merge pull request #6 from Thomas1034/1.20.1-brewing-fuels

Updated comments and moved get() into variable

Jonathing commented:
Closing as stale. I like the idea, but it hasn't been updated in a while and also hasn't been requested of anyone else.

We discussed this on Discord before, but for this PR to move forward, it would need to be effectively rewritten in a way to use data instead of a hard-coded registry event.
